keps/sig-api-machinery/2855-automated-stored-versions/README.md

```diff
@@ -0,0 +1,567 @@
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The [kube-storage-version-migrator] supports migrating existing resources to a new storage version but it does not modify a CRD's `status.storedVersions` field post successful migration. Instead, the cluster administrator is responsible for manually modifying the CRD's `status.storedVersions`. The amount of work this design decision generates for Cluster Administrators will only increase as CRDs are adopted and new versions emerge.
+
+Historically, the [kube-storage-version-migrator] has not managed modifying the `status.storedVersion` field because the CRD's storedVersion could change mid-migration unbenownst to the migrator due to slow/failed watches. 
+
+This KEP proposes a solution that would allow the [kube-storage-version-migrator] to safely modify a CRD's `status.storedVersions` after a successful migration.
+
+[kube-storage-version-migrator]: https://github.com/kubernetes-sigs/kube-storage-version-migrator
+
+## Motivation
+
+It is desirable for the [kube-storage-version-migrator] to modify a CRD's `status.storedVersions` array after a successful migration rather than relying on a cluster admin to manually perform the same steps.
+
+### Goals
+
+- Allow the [kube-storage-version-migrator] to update a CRD's `status.storedVersions`  with the list of stored versions after a successful migration.
+- Ensure that the kube-storage-version-migrator has successfully done its job before removing a stored version from the status of the CRD.
+
+### Non-Goals
+
+- Modify a CRD's `status.storedVersions` field if the migration is not successful.
+
+## Proposal
+
+In order to safely modify a CRD's `status.storedVersions` field, the [kube-storage-version-migrator] would need to prevent changes to the CRD during a migration. CRD's could be marked as migrating using an annotation and changes to the CRD could be prevented by means of an [Validating Admission Webhook](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/).
+
+The validating admission webhook would need to:
+
+- Intercept all Update and Patch operations against CRDs that include the `ksvm.k8s.io/migrating=true` annotation.
+- Allow `/status` requests from the [kube-storage-version-migrator] service account.
+- Allow setting or removing the migrating annotation by the [kube-storage-version-migrator] service account.
+- Prevent other changes to the CRD.
+
+A summary of the new migration workflow can be found below:
+
+1. The [kube-storage-version-migrator] notices a new stored version for the `foo` CRD.
```

Comment:
I don't see why a new storage version is a sufficient reason to start a heavy migration operation. Having a new storage version and old versions in etcd is a totally fine state.

Note that migration can take tens of minutes with enough objects.

Reply:
Ack. Was not aware that ksvm has that trigger controller.

Reply:
Updated the KEP to mention the trigger controller.